Allow a `--print` flag for dashboard URL

ElliotFriend opened the pull request:
Issue #111:

In case the Terminus is being run in a Linux environment, the command
`open http://example.com` will fail.

The `--print` flag will allow the URL to simply be printed to STDOUT
and the user then can copy/paste or do whatever s/he pleases.

Signed-off-by: Elliot Voris elliot@voris.me

ElliotFriend committed:
Issue #111: Allow a `--print` flag for dashboard URL

In case the Terminus is being run in a Linux environment, the command
`open http://example.com` will fail.

The `--print` flag will allow the URL to simply be printed to STDOUT
and the user then can copy/paste or do whatever s/he pleases.

Signed-off-by: Elliot Voris <elliot@voris.me>

ElliotFriend commented:
I'd still like to try and figure out how to utilize `xdg-open` in a Linux environment, but wanted to get the PR out there to start.

ElliotFriend committed:
Issue #111: Check host OS before opening URL

I've put in a `switch` statement so that the command for opening
the URL is different, depending on which host operating system
we're running on.

The command needs to be different for each OS, and now we'll be
using `open` for Mac OS, `xdg-open` for Linux, and `start` for
Windows.

I've not been able to test this (yet) on Mac OS. I can do that
later on tonight, but it would be helpful if somebody else could
test that, too.

Signed-off-by: Elliot Voris <elliot@voris.me>

ElliotFriend commented:
I was able to test that last commit on Linux and Windows, but haven't (yet) been able to test on Mac OS. I can do that later tonight, but feel free to beat me to the punch.

ElliotFriend committed:
Changed the Input::env label to look a little nicer

Previously, the prompt was "Select Choose environment". I changed
that to simply say "Choose environment". It seems a little more
natural, that way.

Signed-off-by: Elliot Voris <elliot@voris.me>

ElliotFriend committed:
Issue #120: Allows the dashboard command to take an optional env flag

If the `--env` flag is specified on the command line, the url will
include that at the end.

If the `--env` flag is not speicified, the command continues to act
as it has. The user will not be prompted to choose an environment,
and the URL will not incoporate any default information.

Signed-off-by: Elliot Voris <elliot@voris.me>

mikevanwinkle commented on php/commands/site.php:
```diff
@@ -242,13 +242,34 @@ public function connection_mode($args, $assoc_args) {
-    Terminus::confirm("Do you want to open your dashboard link in a web browser?", Terminus::get_config());
-    $command = sprintf("open 'https://dashboard.getpantheon.com/sites/%s'", $site->getId());
-    exec($command);
+    $env = @$assoc_args['env'] ? sprintf("#%s", $assoc_args['env']) : null;
```
You should use the Terminus\Helpers\Input function for env() here `Input::env($assoc_args,'env');`

mikevanwinkle replied:
@ElliotFriend can you please update ... also I can test on a mac.

mikevanwinkle commented:
@ElliotFriend so this works on Mac as expected ... however I'm not sure xdg-open is a cureall because it's still environment specific. When I run on ubuntu 14.x I get `sh: 1: xdg-open: not found`

ElliotFriend commented:
I can revisit this a little later tonight, after the kids have gone to bed.

I'm bummed to hear about Ubuntu not having xdg-open. My (most probably inaccurate) understanding was that it was a pseudo-universal tool in most Linux distributions. You mentioned in #111 that you were using a VM, any chance that it's an Ubuntu server install? I wonder if xdg-utils are a part of the desktop distro only?

ElliotFriend commented:
I had briefly played with the `Input::env()` function, but had trouble making it an optional command. So, I went a little cowboy and came up with what you see there. Sitting down now to work with it.

Thanks for the pointer, @mikevanwinkle

ElliotFriend commented:
I just noticed that `Input::env()` doesn't do any checking for whether or not someone submitted dev/test/live as their choice. At first glance, I assumed that the function should return an error if someone submitted `--env=fake`, for instance. But then I realized that multidev would suffer immediately if that were the case.

So, comment withdrawn :smile:

ElliotFriend committed:
Issue #120: Use Input helper function to get the environment

I couldn't figure out how to make `Input::env()` be optional.

It's quite possible that I'm just a putz.

Signed-off-by: Elliot Voris <elliot@voris.me>

ElliotFriend commented:
I used `Input::optional()`, because I couldn't figure out (yet) how to make `Input::env()` not interrupt the command when `--env` wasn't specified.

Thoughts?

ElliotFriend committed:
That space shouldn't be there. My bad.

Signed-off-by: Elliot Voris <elliot@voris.me>

mikevanwinkle commented:
Ok that's cool, optional is good, just want to try to do all input
validation within the Input\Helpers class to make things as readable as
possible. Feel free to modify/add helpers if you need to. Though we need to
find a good way to test them to ensure changes don't break existing
expectations. Trouble is that the blocking IO makes phpunit a problem. I
need to see if there are existing solutions out there we can adapt.

On Wed, Feb 11, 2015 at 6:32 PM, Elliot Voris notifications@github.com
wrote:

> I used Input::optional(), because I couldn't figure out (yet) how to make
> Input::env() not interrupt the command when --env wasn't specified.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pantheon-systems/cli/pull/119#issuecomment-74007316.

## 

Mike Van Winkle
708-289-3136
mike@mikevanwinkle.com
http://www.mikevanwinkle.com
http://www.twitter.com/mpvanwinkle
http://www.facebook.com/mpvanwinkle

"All excellent things are as difficult as they are rare." -Spinoza

ElliotFriend commented:
Definitely seems like a best practice to use those helpers, and I'll definitely keep it in mind moving forward.

ElliotFriend commented:
I just popped in a live dvd of Ubuntu 14.04.1 in, and `xdg-open` seems to be working properly. Although, it's quite a bit more verbose than it is on Linux Mint.

ElliotFriend commented:
Thanks for the merge!
